docs/cli/testing-url-rewrites.md

```diff
@@ -0,0 +1,9 @@
+# Testing URL Rewrites Guide
```

Comment:
I think these tests could actually be an integration test in the CLI right? Just a regular Jest test to be clear, but one that spins up the dev server and hits these endpoints.

This doc as is doesn't really belong here tbh, it should probably be something closer to a code comment or such, but again, a test would be preferable :grin:

Reply:
Agree on moving this to an integration test. 

Took at stab at it, but not getting very far. Do you have any examples that I could go off of? Not sure what the recommended path for jest is to run two processes at once.

Reply:
I'm thinking a very high-level test that serves a very simple app. I think there are two main ways you could go about it. Either rely on `mock-fs` to create a mocked app package with a simple `src/index.ts` and `public/index.html` and whatever other files might be needed. Or you could use a `__fixture__` with an app package that's stored in the repo directly, although it's a bit tricky to be as flexible and try different cases with that approach.

I'm thinking that it's not worth spawning child processes, but instead rely on the high level internal app serve method in the CLI, since without that you can't use `mock-fs` either. You can use `get-port` to find a free local port and put that in config, and then try a couple of `fetch`es towards `localhost:${port}`. Testing towards the real example app in this repo I feel would be too heavyweight, takes too long for it to boot.

Reply:
Alright, I'll take a look. Thanks for the direction!

Reply:
Running into a `jest` brick wall, any help would be appreciated.

My basic idea was to test `await startFrontend(...)`, with a mocked filesystem. I can't mock the full node_modules folder due to a weird symlink recursive thing with techdocs-cli-embedded-app, 
```
ENAMETOOLONG: name too long, stat '/${homedir}/git/backstage/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/node_modules/techdocs-cli-embedded-app/dist/static/react-syntax-highlighter/lowlight-import.c41766b6.chunk.js'

      at ../../../node_modules/mock-fs/lib/loader.js:86:22
          at Array.forEach (<anonymous>)
```
so I am trying to mock each import individually. 

This works up until I need to import `@pmmmwh/react-refresh-webpack-plugin` which has compiled logic to pull `entry: require.resolve('../../client/ErrorOverlayEntry')`. At this point the jest resolver + mock-fs throw an error,
```
    Cannot find module '../../client/ErrorOverlayEntry' from '../../../node_modules/@pmmmwh/react-refresh-webpack-plugin/lib/utils/normalizeOptions.js'
```

Full test case:
```
describe('startFrontend', () => {
  afterEach(mock.restore);
  it('should handle cli options', async () => {
    const getFileFromFileSystem = (name: string) =>
      mock.load(path.join(__dirname, `../../../../${name}`));
    const currentDirectory = process.cwd().split('/')[
      process.cwd().split('/').length - 1
    ];
    const parentDirectory = process.cwd().split('/')[
      process.cwd().split('/').length - 2
    ];
    // jest needs this to enable console.log with mocked full filesystem without node_modules.
    console.log('test');

    mock(
      {
        [path.join(process.cwd(), '../..')]: {
          'package.json': getFileFromFileSystem('package.json'),
          'yarn.lock': getFileFromFileSystem('yarn.lock'),
          'app-config.yaml': `
        app:
            baseUrl: http://localhost:3000/test
        backend:
            baseUrl: http://localhost:7007
        `,
          [parentDirectory]: {
            [currentDirectory]: mock.load(path.join(__dirname, 'test-app')),
          },
          node_modules: [
            '@svgr/webpack',
            'swc-loader',
            'yml-loader',
            'style-loader',
            'css-loader',
            '@pmmmwh/react-refresh-webpack-plugin',
          ].reduce((map, moduleName) => {
            console.log(map);
            const file = getFileFromFileSystem(`node_modules/${moduleName}`);
            if (moduleName.includes('/')) {
              const filePath = moduleName.split('/');
              const scopeDirectory = filePath[filePath.length - 2];
              const mainDirectory = filePath[filePath.length - 1];
              return {
                ...map,
                [scopeDirectory]: {
                  [mainDirectory]: file,
                },
              };
            }
            return {
              ...map,
              [moduleName]: file,
            };
          }, {}),
        },
      },
      { createCwd: false },
    );

    await startFrontend({
      entry: 'src/index',
      verifyVersions: false,
      configPaths: [],
      checksEnabled: false,
    });

    expect(0).toBe(1);
  });
});
```

Reply:
Avoid mocking any of the modules and just focus on the app itself. Tbh I think it's best to set it all up in a fresh mocked project at a path like `/app/...` rather than merging in any existing files in this repo.

If you run into trouble with things being lazy-loaded you can `require()` them before calling `mockFs`, that way they will be in the require cache later on.

The `techdocs-cli-embedded-app` loop seemed like a Yarn bug, Freben ran into that as well. I'd say nuke `node_modules` and re-run `yarn install` if that keeps causing trouble

Reply:
Just pushed what I have for the moment, I ended up sticking with the two processes as the primary testing method for a couple of reasons,
1. The `techdocs-cli-embedded-app` bug kept popping up even after nuking `node_modules`. Was only able to stop it after manually deleting the `node_modules` folder in the `techdocs-cli-embedded-app` symlink.
2. The startFrontend method blocks infinitely and I couldn't find an easy way to read the current process stdout to find an event to trigger an HTTP request (like "webpack compilation completed").

What I have now should be testing the build and local dev server functionality, but doesn't touch the `backendUrl` as I couldn't think of an easy way outside of cypress to test that.